gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/util/DefaultTraversal.java

```diff
@@ -193,7 +193,9 @@ public void applyStrategies() throws IllegalStateException {
     @Override
     public boolean hasNext() {
         if (!this.locked) this.applyStrategies();
-        return this.lastTraverser.bulk() > 0L || this.finalEndStep.hasNext();
+        final boolean more = this.lastTraverser.bulk() > 0L || this.finalEndStep.hasNext();
+        if (!more) CloseableIterator.closeIterator(this);
```

Comment:
Just a comprehension question: Why does it make sense to close the iterator for `hasNext()`, but not do the same for the other terminal steps?

Reply:
the other terminal steps tend to call `fill()` which performs a close properly.